packages/runtime/container-runtime/src/pendingStateManager.ts

```diff
@@ -456,29 +501,43 @@ export class PendingStateManager implements IDisposable {
 			0xa21 /* No pending message found as we start processing this remote batch */,
 		);
 
-		// Note: This could be undefined if this batch became empty on resubmit.
-		// In this case the next pending message is an empty batch marker.
-		// Empty batches became empty on Resubmit, and submit them and track them in case
-		// a different fork of this container also submitted the same batch (and it may not be empty for that fork).
-		const firstMessage = batch.messages.length > 0 ? batch.messages[0] : undefined;
-		const expectedPendingBatchLength = batch.messages.length === 0 ? 1 : batch.messages.length;
+		// If this batch became empty on resubmit, batch.messages will be empty (so firstMessage undefined)
+		// and the next pending message should be an empty batch marker.
+		// More Info: We must submit empty batches and track them in case a different fork
+		// of this container also submitted the same batch (and it may not be empty for that fork).
+		const firstMessage = batch.messages[0];
+		const expectedPendingBatchLength =
+			pendingMessage.batchInfo.length === -1
+				? -1 // Ignore the actual incoming length; -1 length is for back compat so force this to match
+				: batch.messages.length === 0
+					? 1 // For an empty batch, expect a singleton array with the empty batch marker
+					: batch.messages.length; // Otherwise, the lengths should actually match
+
+		// Note: We don't need to use getEffectiveBatchId here, just check the explicit stamped batchID
```

Comment:
Why do we need to compute batchId if we have csn and clientId in the InboundBatch? I know we already talked about this but it's been a minute. I still find unnecessary and confusing having the same data duplicated